SD (and more!) cleanup

tomeichlersmith opened the pull request:
This resolves #39 , #36 , and _un-does_ #55 at the moment. I am working on updating this branch that has gone stale in order to re-align with the current trunk.

### Why?
Dynamically loading SDs will make it much easier to specialize SDs for specific use cases in the future. I think this is a _necessary_ code clean step before we can start effectively wrestling with updating the simulation for the DR geometry.

tomeichlersmith commented on CMakeLists.txt:
```diff
@@ -68,56 +70,24 @@ setup_library(module SimCore
-  #   also used when configuring the makePath.py module
-  set(DB_VERTEX_LIB_INSTALL ${CMAKE_INSTALL_PREFIX}/data/db_vertex_libraries)
-  if(NOT EXISTS ${DB_VERTEX_LIB_INSTALL})
-      #assume new build ==> unpack vertex libraries
-      message(STATUS "Unpacking example dark brem vertex libraries.")
-      file(MAKE_DIRECTORY ${DB_VERTEX_LIB_INSTALL})
-      foreach(library_path ${vertex_libraries})
-          execute_process(COMMAND 
-              tar -C ${DB_VERTEX_LIB_INSTALL} --skip-old-files -zxf ${library_path})
-      endforeach()
-  endif()
-  
-  # Add compiled tests
-  configure_file(${PROJECT_SOURCE_DIR}/test/DarkBremProcessTest.cxx.in 
-                 ${PROJECT_SOURCE_DIR}/test/DarkBremProcessTest.cxx)
-  setup_test(dependencies SimCore::SimCore Geant4::Interface)
-  
-  # Configure the executable so it knows the installation path of the db vertex libraries
-  configure_file(${PROJECT_SOURCE_DIR}/src/SimCore/DarkBrem/print_dark_brem_xsec_table.cxx.in
-                 ${CMAKE_CURRENT_BINARY_DIR}/bin/print_dark_brem_xsec_table.cxx)
-  add_executable(print-dark-brem-xsec-table 
-                 ${CMAKE_CURRENT_BINARY_DIR}/bin/print_dark_brem_xsec_table.cxx)
-  
-  # Link to the Framework library
-  target_link_libraries(print-dark-brem-xsec-table 
-                        PRIVATE Geant4::Interface 
-                                SimCore::SimCore 
-                                Framework::Framework)
-  
-  # Install the fire executable
-  install(TARGETS print-dark-brem-xsec-table 
-          DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
-endif()
-
 setup_python(package_name ${PYTHON_PACKAGE_NAME}/SimCore)
 
-# add visualization executable
-add_executable(g4-vis ${PROJECT_SOURCE_DIR}/src/SimCore/g4_vis.cxx)
-target_link_libraries(g4-vis PRIVATE Geant4::Interface SimCore::SimCore)
-install(TARGETS g4-vis DESTINATION ${CMAKE_INSTALL_PREFIX}/bin)
```
Don't delete these helpful stuff.

tomeichlersmith commented on include/SimCore/LHE/LHEEvent.h:
```diff
@@ -91,12 +92,6 @@ class LHEEvent {
    */
   const double* getVertex() const;
 
-  /**
-   * Get the vertex time
-   * @return time of primary particle creation.
-   */
-  const double getVertexTime() const;
-
```
Don't delete this added functionality.

tomeichlersmith commented on include/SimCore/PluginFactory.h:
```diff
@@ -4,33 +4,36 @@
 #include <variant>
```
This is getting bloated. I think I will replace this with a templated factory as done [in fire](https://github.com/LDMX-Software/fire/blob/trunk/include/fire/factory/Factory.h). I may need to add a vector holding the created objects so many different locations can access the list of manufactured objects.

omar-moreno commented:
How do you handle un-doing #55? That was put there to support LCDD so I'm assuming that's no longer possible?

tomeichlersmith commented:
Yes - I was not planning on keeping the un-doing of #55 , I just wanted to commit something that was working. I am hoping to pull out the necessary changes to AuxInfoReader and then I won't need to undo #55 .

tomeichlersmith committed:
move LHE files into submodule to be compiled separately

this commit WILL NOT compile since it does not include the necessary CMakeLists.txt changes

tomeichlersmith committed:
rm SD creation from AuxInfoReader

again, commit won't compile

tomeichlersmith committed:
move g4user interactions into submodule

won't compile without future CMake changes

tomeichlersmith committed:
add collection names to SimCalorimeterHit as statics

include a clang-formatting of the source files as well

tomeichlersmith committed:
update parallel world to go along with geo move

tomeichlersmith committed:
put LHE submodule in its own namespace

tomeichlersmith committed:
move primary generators to their own submodule

tomeichlersmith committed:
keep trunk updates in PG action

but update to use const parameters instead

tomeichlersmith committed:
updates to PluginFactory (mainly using const Parameters&)

tomeichlersmith committed:
use new header location in run manager

tomeichlersmith committed:
fix up const Parameters in XsecBiasingOp

tomeichlersmith committed:
rm hit writing from persistency manager

it will be done in the SDs themsevles

tomeichlersmith committed:
introduce prototype of dynamically-loaded SD

tomeichlersmith committed:
delete old intermediary G4 hit objects

tomeichlersmith committed:
custom SDs for different subsystems and purposes

tomeichlersmith committed:
add SD parameters to simulator

tomeichlersmith committed:
try out different SD using Scoring planes

tomeichlersmith committed:
reintroduce Geo and merge CMake changes

tomeichlersmith commented on include/SimCore/SDs/EcalSD.h:
```diff
@@ -70,8 +97,8 @@ class EcalSD : public CalorimeterSD {
    */
   std::map<G4VSolid*, G4Polyhedron*> polyMap_;
 
-  /// ConditionsInterface
-  ConditionsInterface& conditionsIntf_;
```
EcalSD and HcalSD need ConditionsInterface so they can access the Geometry conditions for those two subsystems.

tomeichlersmith commented on src/SimCore/SDs/HcalSD.cxx:
```diff
@@ -139,16 +122,15 @@ G4bool HcalSD::ProcessHits(G4Step* aStep, G4TouchableHistory* ROhist) {
+  // Odd layers have bars horizontal
+  // Even layers have bars vertical
   // 5cm wide bars are HARD-CODED
-  if (section == ldmx::HcalID::BACK) {
-    if (hcalGeometry.layerIsHorizontal(layer)) {
-      stripID = int((localPosition.y() + scint->GetYHalfLength()) / 50.0);
-    } else {
-      stripID = int((localPosition.x() + scint->GetXHalfLength()) / 50.0);
-    }
-  } else {
```
Perfect example of why conditions interface is necessary for the *Cal SDs

tomeichlersmith commented on src/SimCore/Simulator.cxx:
```diff
@@ -172,56 +186,44 @@ void Simulator::beforeNewRun(ldmx::RunHeader& header) {
-        G4Electron::Electron()->GetProcessManager()->GetProcessList();
-    int n_electron_processes = electron_processes->size();
-    for (int i_process = 0; i_process < n_electron_processes; i_process++) {
-      G4VProcess* process = (*electron_processes)[i_process];
-      if (process->GetProcessName().contains(
-              darkbrem::G4eDarkBremsstrahlung::PROCESS_NAME)) {
-        // reset process to wrapped process if it is biased
-        if (dynamic_cast<G4BiasingProcessInterface*>(process))
-          process = dynamic_cast<G4BiasingProcessInterface*>(process)
-                        ->GetWrappedProcess();
-        // record the process configuration to the run header
-        dynamic_cast<darkbrem::G4eDarkBremsstrahlung*>(process)->RecordConfig(
-            header);
-        break;
-      }  // this process is the dark brem process
-    }    // loop through electron processes
-  }      // dark brem has been enabled
+  auto apMass{parameters_.getParameter<double>("APrimeMass")};
+  if (apMass > 0) {
+    header.setFloatParameter("A' Mass [MeV]", apMass);
+    header.setFloatParameter(
+        "Dark Brem Global Bias",
+        parameters_.getParameter<double>("darkbrem_globalxsecfactor"));
+    header.setStringParameter(
+        "Dark Brem Vertex Library Path",
+        parameters_.getParameter<std::string>("darkbrem_madgraphfilepath"));
+    header.setIntParameter("Dark Brem Interpretation Method",
+                           parameters_.getParameter<int>("darkbrem_method"));
+  }
```
Undo this reversion.

tomeichlersmith commented on src/SimCore/Simulator.cxx:
```diff
@@ -172,56 +186,44 @@ void Simulator::beforeNewRun(ldmx::RunHeader& header) {
-    if (className.find("simcore::ParticleGun") != std::string::npos) {
+    if (className.find("ParticleGun") != std::string::npos) {
       header.setFloatParameter(genID + " Time [ns]",
                                gen.getParameter<double>("time"));
-      header.setFloatParameter(genID + " Energy [GeV]",
+      header.setFloatParameter(genID + " Energy [MeV]",
```
Undo this reversion to a bug.

Side Note: move all run header config dumping to the generators themselves.

tomeichlersmith reviewed (commented):
Still a lot more to do...

- [x] Updated templated factory design
- [x] Move PrimaryGeneratorAction to G4User?
- [x] Remove ConditionsInterface handle from detector parser, only necessary to be passed to SDs
- [x] Manual comparison of SD changes, make sure they function the same
- [x] Ecal Sim Hit merging so contribs list don't over inflat
- [x] Remove RootPersistencyManager? Not really needed anymore...

tomeichlersmith committed:
transition to newer factory design

tomeichlersmith committed:
move primary generator action to G4user

it is a Geant4 user action, so it should be included in that submodule

tomeichlersmith committed:
move recording of generator configs to their classes

tomeichlersmith committed:
delete root persistency manager

its not needed and the small tasks it does can easily be put into simulator now

tomeichlersmith committed:
update EcalSD to do contrib merging

tomeichlersmith committed:
update HcalSD to use HcalGeometry condition

tomeichlersmith committed:
compiling but then segfaulting on closing of first event

probably missing some initialization that was lost when attempting to
rebase

tomeichlersmith committed:
need to save run number for the sim run for later

the getRunNumber() function calls Process::getRunNumber() which uses the
EventHeader if there is one. Unfortunately, the EventHeader during
onFileClose is not initializaed so the run number is -1, the simple
workaround is to store the run number from onNewRun and use it in
onFileClose

also introduce a commented-out SD::RecordConfig method

tomeichlersmith committed:
allow Factory users to decide on type of pointer

the SensitiveDetectors are deleted by the SD manager so they should be
created into raw pointers while the others can be smart pointers

tomeichlersmith committed:
need HcalGeometry condition for basic test as well

tomeichlersmith committed:
separate TS SDs into the three modules

this will make validation (hopefully) easier so we can make sure these
changes are not altering the output

tomeichlersmith commented:
Things are compiling and running well, now onto the arduous task of validating that these changes that are supposed to change any of the physics don't.

_Edit_: Needs a small change to DecayChildrenKeeper to compile, which makes me think that this _probably_ isn't working as intended since it uses the particle map without the ancestry being fully traced.
```diff
diff --git a/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx b/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx
index a52c7ca..e1c8d81 100644
--- a/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx
+++ b/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx
@@ -10,7 +10,6 @@
 /*   SimCore   */
 /*~~~~~~~~~~~~~*/
 #include "SimCore/UserTrackInformation.h"
-#include "SimCore/UserTrackingAction.h"
 
 namespace biasing {
 namespace utility {
@@ -24,8 +23,7 @@ DecayChildrenKeeper::DecayChildrenKeeper(const std::string& name,
 DecayChildrenKeeper::~DecayChildrenKeeper() {}
 
 void DecayChildrenKeeper::PostUserTrackingAction(const G4Track* track) {
-  const auto& particle_map{
-    simcore::UserTrackingAction::getUserTrackingAction()->getTrackMap()->getParticleMap()};
+  const auto& particle_map{getCurrentParticleMap()};
   if (particle_map.find(track->GetParentID()) != particle_map.end()) {
     const int& parent_pdg{particle_map.at(track->GetParentID()).getPdgID()};
     for (const int& parent : parents_) {

```

_Edit 2_: Nevermind about DecayChildrenKeeper being broken. It is merely checking the PDG if a parent is being saved. It does not rely on the ParticleMap's ancestry using G4Track::GetParentID instead.

tomeichlersmith committed:
update track map to use references instead of pointers

all-around safer and opens door for better const handling

tomeichlersmith committed:
redesign hit merging to more directly align with previous EcalHitIO

tomeichlersmith commented:
Using `SimCore/test/basic.py` as a first testing and my[ tree-diff program from root-diff](https://github.com/tomeichlersmith/root-diff) to deduce different branches.

Output:
```
LDMX_Events mismatched between files
== Branches Only in 'ldmx-sw/trunk/justSim_10_events.root' ==
TargetSimHits_test.id_
TargetSimHits_test.edep_
TargetSimHits_test.x_
TargetSimHits_test.y_
TargetSimHits_test.z_
TargetSimHits_test.time_
TargetSimHits_test.trackIDContribs_
TargetSimHits_test.incidentIDContribs_
TargetSimHits_test.pdgCodeContribs_
TargetSimHits_test.edepContribs_
TargetSimHits_test.timeContribs_
TargetSimHits_test.nContribs_
TrackerScoringPlaneHits_test.id_
TrackerScoringPlaneHits_test.layerID_
TrackerScoringPlaneHits_test.moduleID_
TrackerScoringPlaneHits_test.edep_
TrackerScoringPlaneHits_test.time_
TrackerScoringPlaneHits_test.px_
TrackerScoringPlaneHits_test.py_
TrackerScoringPlaneHits_test.pz_
TrackerScoringPlaneHits_test.energy_
TrackerScoringPlaneHits_test.x_
TrackerScoringPlaneHits_test.y_
TrackerScoringPlaneHits_test.z_
TrackerScoringPlaneHits_test.pathLength_
TrackerScoringPlaneHits_test.trackID_
TrackerScoringPlaneHits_test.pdgID_
== Branches with different content ==
EcalSimHits_test.x_
EcalSimHits_test.y_
EcalSimHits_test.z_
EventHeader.timestamp_.fSec
EventHeader.timestamp_.fNanoSec
```

- Ignore timestamp difference
- [x] Look into positional difference, do we need to keep `EcalSD::getHitPosition`? It doesn't seem necessary to me...
- [x] Include TrackerScoringPlanes in default SD config
- [x] Include a TargetSD in default SD config

tomeichlersmith committed:
add recoil tracker scoring planes to default SD config

tomeichlersmith committed:
fix to position determination to align with current

we serialize the cell-center position rather than the actual simulated
position

tomeichlersmith committed:
use TrigScintSD for target sim hits

this is what is currently done on trunk as well so it produces pretty
much identical results

tomeichlersmith commented:
Alright, I've fixed up the differences except for the incident ID contribs for the TargetSimHits. It looks like the new design is catching another hit that has a non-primary incident particle where-as before it was just going up to the default primary.

**Note** I'm only comparing **ten** events right now. I will update with a larger 10k run.

@omar-moreno Do you think these are close enough to not care?
@bryngemark Do y'all use the incident ID contribs for the TargetSimHits in any TS analyses?

![target-sim-hits-incident-id-contribs](https://user-images.githubusercontent.com/31970302/182918224-111a64d8-9e04-4912-8768-bd38654289d4.png)

tomeichlersmith committed:
documentation updates

bryngemark commented:
> @bryngemark Do y'all use the incident ID contribs for the TargetSimHits in any TS analyses?

No, not that I'm aware of. (We're not really using contribs at all at the moment.) So far we've relied on truth info in the TS itself, not in the target.

tomeichlersmith commented:
Cool, I'm going to make larger samples to make sure that is the only part changing, and if it is then I will call it validated :+1:

tomeichlersmith committed:
need an EndOfEvent for aborted events

we need a call back so the SDs can clean up their hit collections when
an event ends, this was only caught when attempting biasing where events
are aborted many times before a successful event is serialized

tomeichlersmith committed:
patch process name in SimParticle

the update that occurred on trunk was lost during a rebase, simple fix

tomeichlersmith commented:
# Validation of SD Redesign
Comparison of sim-level information between `trunk` and `39-sd-cleanup` of SimCore.

## trunk
These samples were generated with the production container built from the latest
commit to trunk `1c446d2`
```bash
ldmx pull pro sha-1c446d24
```

## 39-sd-cleanup
These samples were generated with a local build using dev/latest.

## Comparison
The comparison was done using the `tree-diff` executable from 
[tomeichlersmith/root-diff repository](https://github.com/tomeichlersmith/root-diff).
This lists the branches with different content on a _bit-by-bit_ level.
These developments are not intended to change the simulated physics _at all_,
so I am working towards making the files differ _only_ in their time stamps.

## Sample Configs
They are listed with respect to ldmx-sw. All of these were run with the developments in SimCore
and the attached updates to ldmx-sw/Biasing and match files generated with `trunk` as described
above. Some did have slightly different incident ID contribs for the Target/TriggerPad sim hits.
Downstream usage of these sim hits do not use the incident ID because that was designed for the
calorimeters, so that difference is ignored.
- SimCore/test/basic.py 10 and 10k events
- Biasing/test/ecal_pn.py 1k events
- Biasing/test/ecal_db.py 100 events
- Biasing/test/energy_sort.py 10 events
- Biasing/test/target_db.py 100 events
- Biasing/test/target_en.py 1k events
- Biasing/test/target_mumu.py 1k events
- Biasing/test/target_pn.py 1k events

## ldmx-sw Changes
Below is a `git diff` of what needs to be changes in ldmx-sw/Biasing after these changes.
All of the changes are simply updating the Biasing/test configs that had gone stale except
the compilation-necessary change in DecayChildrenKeeper.

```diff
diff --git a/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx b/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx
index a52c7ca..6d8268e 100644
--- a/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx
+++ b/Biasing/src/Biasing/Utility/DecayChildrenKeeper.cxx
@@ -10,7 +10,7 @@
 /*   SimCore   */
 /*~~~~~~~~~~~~~*/
 #include "SimCore/UserTrackInformation.h"
-#include "SimCore/UserTrackingAction.h"
+#include "SimCore/G4User/TrackingAction.h"
 
 namespace biasing {
 namespace utility {
@@ -24,8 +24,7 @@ DecayChildrenKeeper::DecayChildrenKeeper(const std::string& name,
 DecayChildrenKeeper::~DecayChildrenKeeper() {}
 
 void DecayChildrenKeeper::PostUserTrackingAction(const G4Track* track) {
-  const auto& particle_map{
-    simcore::UserTrackingAction::getUserTrackingAction()->getTrackMap()->getParticleMap()};
+  const auto& particle_map{simcore::g4user::TrackingAction::get()->getTrackMap().getParticleMap()};
   if (particle_map.find(track->GetParentID()) != particle_map.end()) {
     const int& parent_pdg{particle_map.at(track->GetParentID()).getPdgID()};
     for (const int& parent : parents_) {
diff --git a/Biasing/test/ecal_db.py b/Biasing/test/ecal_db.py
index dc4c58b..1d128c6 100644
--- a/Biasing/test/ecal_db.py
+++ b/Biasing/test/ecal_db.py
@@ -3,6 +3,7 @@ p = ldmxcfg.Process('ecal_db')
 from LDMX.Biasing import eat
 from LDMX.SimCore import makePath
 from LDMX.Ecal import EcalGeometry
+from LDMX.Hcal import HcalGeometry
 p.sequence = [
     eat.dark_brem( 
         10., #MeV - mass of A'
@@ -10,5 +11,6 @@ p.sequence = [
         'ldmx-det-v12' , #name of geometry to use
         )
     ]
+p.maxTriesPerEvent = 1000
 p.maxEvents = 100
-p.outputFiles = [ '/tmp/ecal_db.root' ]
+p.outputFiles = [ 'ecal_db.root' ]
diff --git a/Biasing/test/ecal_pn.py b/Biasing/test/ecal_pn.py
index f6d51f1..be25175 100644
--- a/Biasing/test/ecal_pn.py
+++ b/Biasing/test/ecal_pn.py
@@ -4,6 +4,7 @@ p = ldmxcfg.Process('ecal_pn')
 from LDMX.Biasing import ecal
 from LDMX.SimCore import generators
 import LDMX.Ecal.EcalGeometry
+import LDMX.Hcal.HcalGeometry
 p.sequence = [
     ecal.photo_nuclear( 
         'ldmx-det-v12' , 
@@ -11,4 +12,5 @@ p.sequence = [
         )
     ]
 p.maxEvents = 1000
-p.outputFiles = [ '/tmp/ecal_pn.root' ]
+p.maxTriesPerEvent = 1000
+p.outputFiles = [ 'ecal_pn.root' ]
diff --git a/Biasing/test/energy_sort.py b/Biasing/test/energy_sort.py
index 0cafe1e..4133621 100644
--- a/Biasing/test/energy_sort.py
+++ b/Biasing/test/energy_sort.py
@@ -1,17 +1,17 @@
 from LDMX.Framework import ldmxcfg
 # create my process object
 p = ldmxcfg.Process( "test" )
+p.run = 9001
 p.maxEvents = 10
-p.outputFiles = [ "/tmp/energy_sort.root" ]
+p.outputFiles = [ "energy_sort.root" ]
 from LDMX.SimCore import simulator as sim
-from LDMX.Ecal import EcalGeometry
+import LDMX.Ecal.EcalGeometry
+import LDMX.Hcal.HcalGeometry
 mySim = sim.simulator( "mySim" )
 mySim.setDetector( 'ldmx-det-v12' )
-mySim.runNumber = 9001
 from LDMX.SimCore import generators as gen
 mySim.generators.append( gen.single_4gev_e_upstream_tagger() )
 mySim.description = 'Basic test Simulation'
-mySim.randomSeeds = [ 1 , 2 ]
 from LDMX.Biasing import util
 mySim.actions = [ util.PartialEnergySorter( 3000. ) ]
 p.sequence.append( mySim )
diff --git a/Biasing/test/target_db.py b/Biasing/test/target_db.py
index 88f81c5..e1cd2a1 100644
--- a/Biasing/test/target_db.py
+++ b/Biasing/test/target_db.py
@@ -2,7 +2,8 @@ from LDMX.Framework import ldmxcfg
 p = ldmxcfg.Process('target_db')
 from LDMX.Biasing import target
 from LDMX.SimCore import makePath
-from LDMX.Ecal import EcalGeometry
+import LDMX.Ecal.EcalGeometry
+import LDMX.Hcal.HcalGeometry
 p.sequence = [
     target.dark_brem( 
         10., #MeV - mass of A'
@@ -11,4 +12,5 @@ p.sequence = [
         )
     ]
 p.maxEvents = 1000
-p.outputFiles = [ '/tmp/target_db.root' ]
+p.maxTriesPerEvent = 1000
+p.outputFiles = [ 'target_db.root' ]
diff --git a/Biasing/test/target_en.py b/Biasing/test/target_en.py
index 6c99345..37f6481 100644
--- a/Biasing/test/target_en.py
+++ b/Biasing/test/target_en.py
@@ -3,6 +3,7 @@ p = ldmxcfg.Process('target_en')
 from LDMX.Biasing import target
 from LDMX.SimCore import generators
 import LDMX.Ecal.EcalGeometry
+import LDMX.Hcal.HcalGeometry
 p.sequence = [
     target.electro_nuclear( 
         'ldmx-det-v12' , 
@@ -10,4 +11,5 @@ p.sequence = [
         )
     ]
 p.maxEvents = 1000
-p.outputFiles = [ '/tmp/target_en.root' ]
+p.maxTriesPerEvent = 1000
+p.outputFiles = [ 'target_en.root' ]
diff --git a/Biasing/test/target_mumu.py b/Biasing/test/target_mumu.py
index cb7d664..6addac0 100644
--- a/Biasing/test/target_mumu.py
+++ b/Biasing/test/target_mumu.py
@@ -3,6 +3,7 @@ p = ldmxcfg.Process('target_mumu')
 from LDMX.Biasing import target
 from LDMX.SimCore import generators
 import LDMX.Ecal.EcalGeometry
+import LDMX.Hcal.HcalGeometry
 p.sequence = [
     target.gamma_mumu(
         'ldmx-det-v12' ,
@@ -10,4 +11,5 @@ p.sequence = [
         )
     ]
 p.maxEvents = 1000
-p.outputFiles = [ '/tmp/target_mumu.root' ]
+p.maxTriesPerEvent = 1000
+p.outputFiles = [ 'target_mumu.root' ]
diff --git a/Biasing/test/target_pn.py b/Biasing/test/target_pn.py
index c8d4fff..dde9e04 100644
--- a/Biasing/test/target_pn.py
+++ b/Biasing/test/target_pn.py
@@ -3,6 +3,7 @@ p = ldmxcfg.Process('target_pn')
 from LDMX.Biasing import target
 from LDMX.SimCore import generators
 import LDMX.Ecal.EcalGeometry
+import LDMX.Hcal.HcalGeometry
 p.sequence = [
     target.photo_nuclear( 
         'ldmx-det-v12' , 
@@ -10,4 +11,5 @@ p.sequence = [
         )
     ]
 p.maxEvents = 1000
-p.outputFiles = [ '/tmp/target_pn.root' ]
+p.maxTriesPerEvent = 1000
+p.outputFiles = [ 'target_pn.root' ]
```

tomeichlersmith commented:
@omar-moreno @bryngemark I feel like this is fully validated, but I want to get confirmation before I merge such a large change. Any final thoughts/concerns?

tomeichlersmith committed:
patch for getCurrentParticleMap to be linked properly

depending on symbol lookup, the static member was not showing, just
going back to normal const member worked well

tomeichlersmith committed:
don't need event handle for cleanup

when having the SDs cleanup on an aborted event, we don't need to pass
them the event handle

tomeichlersmith committed:
document configuration of SDs
